crates/red_knot_python_semantic/src/types/call/bind.rs

```diff
@@ -589,53 +614,91 @@ impl<'db> Bindings<'db> {
+                                params |= DataclassParams::WEAKREF_SLOT;
                             }
 
-                            overload.set_return_type(Type::DataclassDecorator(metadata));
+                            overload.set_return_type(Type::DataclassDecorator(params));
                         }
                     }
 
-                    _ => {}
+                    Some(KnownFunction::DataclassTransform) => {
+                        if let [eq_default, order_default, kw_only_default, frozen_default, _field_specifiers, _kwargs] =
+                            overload.parameter_types()
+                        {
+                            let mut params = DataclassTransformerParams::empty();
+
+                            if to_bool(eq_default, true) {
+                                params |= DataclassTransformerParams::EQ_DEFAULT;
+                            }
+                            if to_bool(order_default, false) {
+                                params |= DataclassTransformerParams::ORDER_DEFAULT;
+                            }
+                            if to_bool(kw_only_default, false) {
+                                params |= DataclassTransformerParams::KW_ONLY_DEFAULT;
+                            }
+                            if to_bool(frozen_default, false) {
+                                params |= DataclassTransformerParams::FROZEN_DEFAULT;
+                            }
+
+                            overload.set_return_type(Type::DataclassTransformer(params));
+                        }
+                    }
+
+                    _ => {
+                        if let Some(params) = function_type.dataclass_transformer_params(db) {
+                            // This is a call to a custom function that was decorated with `@dataclass_transformer`.
+                            // If this function was called with a keyword argument like `order=False`, we extract
+                            // the argument type and overwrite the corresponding flag in `dataclass_params` after
+                            // constructing them from the `dataclass_transformer`-parameter defaults.
+
+                            let mut dataclass_params =
+                                DataclassParams::from_transformer_params(params);
+
+                            if let Some(Some(Type::BooleanLiteral(order))) = callable_signature
+                                .iter()
+                                .nth(overload_index)
+                                .and_then(|signature| {
+                                    let (idx, _) =
+                                        signature.parameters().keyword_by_name("order")?;
+                                    overload.parameter_types().get(idx)
+                                })
+                            {
+                                dataclass_params.set(DataclassParams::ORDER, *order);
+                            }
```

Comment:
Haven't invested more time into making this prettier, as I'm not sure if that one failing test case at the bottom if `dataclass_transform.md` will affect how we handle overloaded dataclass-transformers in general.